cycode/cli/utils/machine_id.py

```diff
@@ -0,0 +1,117 @@
+
+
+def _read_cmd(cmd: str) -> Optional[str]:
+    try:
+        return shell(cmd, execute_in_shell=True)
+    except:  # noqa
+        return None
+
+
+def _read_registry(registry: str, key: str, registry_key: Optional[int] = None) -> Optional[str]:
+    import winreg
+
+    if not registry_key:
+        registry_key = winreg.HKEY_LOCAL_MACHINE
+
+    with winreg.ConnectRegistry(None, registry_key) as reg:
+        with winreg.OpenKey(reg, registry) as key_object:
+            try:
+                value, _ = winreg.QueryValueEx(key_object, key)
+                return value
+            except WindowsError:
+                pass
+
+    return None
+
+
+def _read_file(path: str) -> Optional[str]:
+    try:
+        with open(path, encoding='UTF-8') as f:
+            return f.read()
+    except:  # noqa
+        return None
+
+
+def _get_darwin_machine_id() -> str:
+    return _read_cmd("ioreg -d2 -c IOPlatformExpertDevice | awk -F\\\" '/IOPlatformUUID/{print $(NF-1)}'")
+
+
+def _get_windows_machine_id() -> str:
+    mid = _read_registry('SOFTWARE\\MICROSOFT\\CRYPTOGRAPHY', 'MachineGuid')
+    if not mid:
+        mid = _read_cmd('wmic csproduct get uuid').split('\n')[2].strip()
+
+    return mid
+
+
+def _get_linux_machine_id() -> str:
+    mid = _read_file('/var/lib/dbus/machine-id')
+
+    if not mid:
+        mid = _read_file('/etc/machine-id')
+
+    if not mid:
+        cgroup = _read_file('/proc/self/cgroup')
+        if cgroup and 'docker' in cgroup:
+            mid = _read_cmd('head -1 /proc/self/cgroup | cut -d/ -f3')
+
+    if not mid:
+        mountinfo = _read_file('/proc/self/mountinfo')
+        if mountinfo and 'docker' in mountinfo:
+            mid = _read_cmd("grep 'systemd' /proc/self/mountinfo | cut -d/ -f3")
+
+    return mid
+
+
+def _get_bsd_machine_id() -> str:
+    mid = _read_file('/etc/hostid')
+    if not mid:
+        mid = _read_cmd('kenv -q smbios.system.uuid')
+
+    return mid
+
+
+def _get_machine_id() -> str:
+    if platform == 'darwin':
+        mid = _get_darwin_machine_id()
+    elif platform in {'win32', 'cygwin', 'msys'}:
+        mid = _get_windows_machine_id()
+    elif platform.startswith('linux'):
+        mid = _get_linux_machine_id()
+    elif platform.startswith('openbsd') or platform.startswith('freebsd'):
+        mid = _get_bsd_machine_id()
+    else:
+        raise CycodeError('Unknown platform')
+
+    if mid is None:
+        raise CycodeError("Can't get Machine ID")
+
+    return mid
+
+
+def _get_hmac_hex(key: str, msg: str) -> str:
+    return hmac.new(key=key.encode(), msg=msg.encode(), digestmod=hashlib.sha256).hexdigest()
+
+
+@lru_cache(maxsize=None)
+def machine_id() -> str:
+    return _get_machine_id()
+
+
+@lru_cache(maxsize=None)
+def protected_machine_id(app_id: str) -> str:
+    """Calculates HMAC-SHA256 of the app ID, keyed by the machine ID and returns a hex-encoded str."""
+    return _get_hmac_hex(key=_get_machine_id(), msg=app_id)
```

Comment:
I don't really like collecting so much deep data and running shell here
I suggest to have the following:

"{OS_type}-{python_version}-{uniq_id}"
where uniq_id will be a hex/sha from one of the following options:
- mac address
- installation/home dir
- we'll write some uniq uuiq to `~/.cycode/id` file

Reply:
Here is only one exec/file read. The implementation supports all environments and operation systems. That's why code contains many execs. The purpose of the code is simple: get OS installation ID. 

To get a mac address we should implement something similar. Because the way to retrieve it depends on the OS too. 

The solution with generating cycode cli installation ID instead of using OS installation ID sounds okay and was my second option to machine id

Reply:
do we cache somewhere info about the installation id in the machine?

Reply:
Installation ID of OS? It cached in runtime. 

We don't have any logic for the installation ID of CLI yet

Reply:
why we need the machine id? I agree with @artem-fedorov , lets keep it simple

Reply:
we need some uniq id for build some dashboards how many users of CLI do we have :) 
I already solved similar issue in jmeter-plugins : https://github.com/undera/jmeter-plugins-manager/blob/master/src/main/java/org/jmeterplugins/repository/JARSourceHTTP.java#L304

Reply:
@MarshalX field in config.yaml sounds good to me. @artem-fedorov WDYT?

Reply:
who generate `config.yaml` and how often do we refresh it?

Reply:
we generate it if its needed, for exmaple for adding ignores by value/secret/etc. it can be generated somehow in the first run for adding the id. also we can get another id of the local comp, but not that id that requires lots of logic haha :)

Reply:
> who generate `config.yaml` and how often do we refresh it?

CLI creates if not exist. Refresh? What do you mean. We don't delete this config at all

Reply:
In case this is our internal id, we can save random uuiq in this file